README.md
# Radiotray-NG: An Internet radio player for Linux #

It became clear as one of the early contributors to the RadioTray project that it
was not getting the attention it required and was probably dead. A lot of the 
technologies it used had moved on to newer versions and the bugs started piling up.
I did my best to help users, but a new start was required.

The version here is what "I" wanted out of RadioTray.

## Radiotray-NG goals were: ##

* Cross platform friendly design as I occasionally use a Mac.
* Better gstreamer error handling and recovery.
* Fix RadioTray's flawed bookmark format.
* Build in the only RadioTray plugin that I felt I needed, which is a shutdown timer.
* Per station/group notification icon support.
* Better parsing of the stream meta data and optionally more stream info displayed.
* A bit more attention to notification details and formatting.

## Where Radiotray-NG is: ##

* RadioTray functionality
* Theme support
* Debian packaging
* Full AppIndicator support
* No groups within groups by design to keep the interface clean.
* Volume up/down support using mouse wheel (left/right can also be used)
* Media key support
* Dbus interface for controlling Radiotray-NG and accessing stream meta data
* --play command line option for resuming playback

## Future: ##

* ~~Cross platform bookmark editor application.~~
* Cross compile on Mac (via brew packages) with icons, menu, media keys etc.
* ~~Script to convert RadioTray's bookmarks.xml to the new format.~~

**Icons:** [http://www.iconsplace.com/]
**License:** Creative Commons Attribution-NoDerivs 3.0 Unported

## Bugs: ##

* Radiotray-NG's user-agent is created at compile time. Which means a binary created on 16.10 will report that even when run on 16.04.

## Screenshots: ##

![Screenshot](images/screenshot_1.png)
![Screenshot](images/screenshot_7.png)
![Screenshot](images/screenshot_2.png)
![Screenshot](images/screenshot_3.png)
![Screenshot](images/screenshot_4.png)
![Screenshot](images/screenshot_6.png)
![Screenshot](images/screenshot_5.png)


## Config File ##

A config (radiotray-ng.json) is created in your ~/.config/radiotray-ng directory with the following defaults:
```
{
   "bookmarks" : "~/.config/radiotray-ng/bookmarks.json",
   "bookmark-editor" : "rtng-bookmark-editor",
   "buffer-duration" : 2,
   "buffer-size" : 320000,
   "compact-menu" : false,
   "debug-logging" : false,
   "file-monitor" : true,
   "file-monitor-interval" : 10,
   "invert-menu" : false,
   "last-station" : "",
   "last-station-group" : "",
   "notification-verbose" : true,
   "notifications" : true,
   "sleep-timer" : 30,
   "split-title" : true,
   "track-info-copy" : false,
   "ihr-title" : true,
   "tag-info-verbose" : true,
   "volume-level" : 100,
   "volume-step" : 1,
   "wrap-track-info" : true,
   "wrap-track-info-len" : 40,
   "media-key-mapping" : false,
   "media-key-previous-station" : "Previous",
   "media-key-next-station" : "Next",
   "media-key-volume-up" : "",
   "media-key-volume-down" : "",
   "media-keys-old-dbus-name" : false,
   "radiotray-ng-on" : "radiotray-ng-on",
   "radiotray-ng-off" : "radiotray-ng-off",
   "radiotray-ng-notification" : "radiotray-ng-notification",
   "root-group-bottom-pos" : true
}
```
* **Installed config will only include commonly edited entries, all others are using default values shown above.**
```
                 bookmarks: location of bookmarks file
           bookmark-editor: bookmark editor to launch
               buffer-size: size of buffer in bytes
           buffer-duration: number of seconds to buffer
             compact-menu : enable/disable the use of menu separators
             debug-logging: enable/disable verbose debug logging
              file-monitor: enable/disable notifcation of bookmark file changes
     file-monitor-interval: time in seconds to poll for bookmark file changes
              invert-menu : menu rendered in reverse order
      notification-verbose: more status information than normal
             notifications: turns on/off notification messages
               sleep-timer: value is in minutes
               split-title: attempts to reformat the notification into title/artist
                 ihr-title: extra parsing for iheartradio formatting (requires split-title)
          tag-info-verbose: displays in the menu stream information such as bitrate etc.
           track-info-copy: enable/disable track clicking to copy into clipboard 
               volume-step: value used to increment/decrement the volume level
           wrap-track-info: enable/disable the wrapping of title & artist menu text
       wrap-track-info-len: maximum title & artist line length 
         media-key-mapping: enable the mapping of media keys to volume up/down etc. (Previous, Next, Rewind, FastForward etc.)
media-key-previous-station: media key to use for previous station within current group
    media-key-next-station: media key to use for next station within current group
       media-key-volume-up: media key to use for volume up
     media-key-volume-down: media key to use for volume down
  media-keys-old-dbus-name: force the use of old dbus name
           radiotray-ng-on: installed theme icon name for "on" or path to image
          radiotray-ng-off: installed theme icon name for "off" or path to image
 radiotray-ng-notification: installed theme icon name for "notification" or path to image
     root-group-bottom-pos: render root group stations at bottom or top of menu

```
* Do not edit the config while Radiotray-NG is running or your changes will be lost.
* No checks are made if a media key assignment collides with another action.

## Bookmarks Format ##

Bookmarks are defined in the following JSON format:
```
[
   {
      "group" : "group name 1",
      "image" : null,
      "stations" : [
         {
            "image" : null,
            "name" : "station name",
            "url" : "http://station/station.pls"
         },
         ...
        ]
   },
   {
      "group" : "group name 2",
      "image" : "~/Dropbox/radiotray-ng/images/station.png",
      "stations" : [
         {
            "image" : null,
            "name" : "station name",
            "url" : "http://station/station.pls",
            "notifications" : false
         },
         ...
        ]
   },
   ...
]
```
Group with the name "root" is treated differently and is rendered at the base of the menu. A group's image specifies the notification icon to display and is inherited by all of the stations. A station with an image overrides the group image.

Use the "Preferences/Reload Bookmarks" option to see your changes.


## Convert RadioTray Bookmarks ##

The rt2rtng script will convert your RadioTray bookmarks.xml file. All groups within groups are moved to the root and any empty ones are removed. It's not a perfect conversion and will no doubt require some editing.

```
$ rt2rtng ~/.local/share/radiotray/bookmarks.xml > bookmarks.json
```


## Bookmarks Editor ##

The Bookmarks Editor is a simple editor that enables management of your bookmarks file eliminating the need for manually editing the file. The editor will attempt to open the default bookmarks file on startup, but you can select to open another bookmarks file if desired. This provides you with the ability to manage multiple bookmark files.

The editor supports all of the typical editor operations including adding, editing and deleting of both groups and stations. Images are easily selected via standard browsing functionality. You can also arrange the groups and stations using simple drag-n-drop actions.

* Use Radiotray-NG's reload bookmarks after saving your changes.

## DBus Interface ##

```
Available commands:

    volume_up
    volume_down
    set_volume 'level'
    play
    play_url 'url'
    stop
    quit
    previous_station
    reload_bookmarks
    next_station
    get_bookmarks
    get_config
    get_player_state
    play_station 'group' 'station'
```
```
Example:

$ qdbus com.github.radiotray_ng /com/github/radiotray_ng com.github.radiotray_ng.get_player_state
{
   "artist" : "Suspense",
   "bitrate" : "24 kb/s",
   "codec" : "MPEG-1 Layer 3 (MP3)",
   "group" : "Old Time Radio",
   "image" : "radiotray-ng-notification",
   "state" : "playing",
   "station" : "AM 1710 Antioch OTR",
   "title" : "Jul 15, 1948: Summer Night w/Ida Lupino",
   "volume" : "15"
}
```

## Install ##

Download a release or clone the repo and build the latest Debian package.

https://github.com/ebruck/radiotray-ng/releases


## To Build on Ubuntu: ##

Install these packages:
```
lsb-release libcurl4-openssl-dev libjsoncpp-dev libxdg-basedir-dev libnotify-dev libboost-filesystem-dev libgstreamer1.0-dev libappindicator3-dev libboost-log-dev libboost-program-options-dev libgtk-3-dev libnotify-dev lsb-release libbsd-dev libncurses5-dev libglibmm-2.4-dev libwxgtk3.0-dev libwxgtk3.0-0v5 cmake
```

## Build Radiotray-NG & Debian Package ##
```
$ git clone https://github.com/ebruck/radiotray-ng.git
$ cd radiotray-ng
$ mkdir build
$ cd build
$ cmake .. -DCMAKE_BUILD_TYPE=Release
$ make package
$ sudo dpkg -i ./radiotray-ng_x.y.z_<i386|amd64>.deb
$ sudo apt-get install -f
```

## Build Radiotray-NG + Tests & Debian Package ##
```
$ git clone https://github.com/ebruck/radiotray-ng.git
$ cd radiotray-ng
$ mkdir build
$ cd build
$ cmake .. -DBUILD_TESTS=ON -DCMAKE_BUILD_TYPE=Debug
$ make package
$ sudo dpkg -i ./radiotray-ng_x.y.z_<i386|amd64>.deb
$ sudo apt-get install -f
```

## To Build on Fedora: ##

Install these packages:
```
redhat-lsb cmake libcurl-devel boost-devel wxGTK3-devel jsoncpp-devel gstreamer1-devel libxdg-basedir-devel libbsd-devel libappindicator-gtk3-devel libnotify-devel glibmm24-devel
```
### Thank you JetBrains for providing a free open source license for [CLion](https://www.jetbrains.com/clion/).



debian/changelog
radiotray-ng (0.2.5) unstable; urgency=low

      * Updated & added new stations
      * Added dbus play_url method
      * Added initial m3u8 support
      * Various dbus & editor bug fixes
      * Added rtng-dbus & rtng2xml scripts

 -- Edward G. Bruck <ed.bruck1@gmail.com>


radiotray-ng (0.2.4) unstable; urgency=low

      * Added dbus set_volume method
      * Added root-group-bottom-pos config option
      * Notification icon visibility on light backgrounds improved
      * Minor build changes & bug fixes

 -- Edward G. Bruck <ed.bruck1@gmail.com>  Wed, 17 Oct 2018 21:19:49 -0700


radiotray-ng (0.2.3) unstable; urgency=low

      * Fixed playlist line ending type detection
      * Added wrap-track-info & wrap-track-info-len config options
      * Better handling of text/uri-list & text/html playlists
      * Minor build changes

 -- Edward G. Bruck <ed.bruck1@gmail.com>  Fri, 13 Jul 2018 21:30:00 -0700


radiotray-ng (0.2.2) unstable; urgency=low

      * Added copying title & artist to clipboard config option
      * Added per station notification option & editor support
      * Added dbus reload_bookmarks & quit methods
      * Minor build changes

 -- Edward G. Bruck <ed.bruck1@gmail.com>  Tue, 24 Apr 2018 21:25:51 -0700


radiotray-ng (0.2.1) unstable; urgency=low

      * Fixed editor group & station reordering
      * Added BUILD_TESTS cmake option
      * Added dbus get_config method

 -- Edward G. Bruck <ed.bruck1@gmail.com>  Sat, 17 Feb 2018 08:52:41 -0800


radiotray-ng (0.2.0) unstable; urgency=low

      * Added contributed Bookmark Editor (Thanks Mike!)
      * Added iHeartRadio tag parsing
      * Added buffer-duration config option
      * Fixed jsoncpp build warnings breaking Arch builds
      * Minor bug fixes

 -- Edward G. Bruck <ed.bruck1@gmail.com>  Sat, 30 Dec 2017 16:27:34 -0800


radiotray-ng (0.1.8) unstable; urgency=low

      * Added icon middle click support for starting/stopping playback
      * Added command line help etc.
      * Added invert-menu configuration option for bottom dock users
      * Added Bluegrass station to default bookmark file
      * Fixed wrong api used for menu separators
      * Fixed output not using autoaudiosink
      * Reduced notifications if player buffers

 -- Edward G. Bruck <ed.bruck1@gmail.com>  Sat, 07 Oct 2017 11:00:07 -0700


radiotray-ng (0.1.7) unstable; urgency=low

      * Added dbus interface
      * Added --play option for resuming playback on startup
      * Added optional notification of external bookmark file changes
      * Updated to libappindicator3

 -- Edward G. Bruck <ed.bruck1@gmail.com>  Sat, 02 Sep 2017 11:01:19 -0700


radiotray-ng (0.1.6) unstable; urgency=low

      * new icon config overrides
      * sleep timer dialog added
      * support for gnome-settings-daemon media key api changes
      * added Fedora build instructions

 -- Edward G. Bruck <ed.bruck1@gmail.com>  Wed, 28 Jun 2017 20:16:07 -0700


radiotray-ng (0.1.5) unstable; urgency=low

      * Fixed startup crash seen on some systems using gcc 6.x
      * Added compact-menu option to disable the use of menu separators
      * Removed ncurses requirement if appindicator library is found
      * Notification uses station name if artist tag is missing
      * Smarter boost dependency detection
      * Added rt2rtng script for converting RadioTray bookmarks
      * Fixed user-agent and release build issues
      * Logging & aboutbox now include git commit used to generate binary
      * Added optional mapping of media keys for previous/next station etc

 -- Edward G. Bruck <ed.bruck1@gmail.com>  Sat, 22 Apr 2017 16:04:25 -0700


radiotray-ng (0.1.4) unstable; urgency=low

      * Disabling notifications prevented stopping of player
      * Fixed lintian package errors

 -- Edward G. Bruck <ed.bruck1@gmail.com>  Thu, 23 Feb 2017 18:49:38 -0800


radiotray-ng (0.1.3) unstable; urgency=low

      * Support building on 32bit architectures
      * Added 24x24 icons and pid file to prevent multiple instances from
        running

 -- Edward G. Bruck <ed.bruck1@gmail.com>  Sat, 18 Feb 2017 14:44:27 -0800


radiotray-ng (0.1.2) unstable; urgency=low

      *  Fixed gdk threading not being initialized before calling gtk_init
      *  Don't overwrite bad bookmarks file on startup

 -- Edward G. Bruck <ed.bruck1@gmail.com>  Wed, 15 Feb 2017 19:47:03 -0800


radiotray-ng (0.1.1) unstable; urgency=low

	  * Fixed word_wrap infinite loop

 -- Edward G. Bruck <ed.bruck1@gmail.com>  Sun, 12 Feb 2017 12:58:39 -0800


radiotray-ng (0.1.0) unstable; urgency=low

	  * Initial release.

 -- Edward G. Bruck <ed.bruck1@gmail.com>  Sat, 11 Feb 2017 12:10:00 -0800


include/radiotray-ng/common.hpp
// Copyright 2017 Edward G. Bruck <ed.bruck1@gmail.com>
//
// This file is part of Radiotray-NG.
//
// Radiotray-NG is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// Radiotray-NG is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with Radiotray-NG.  If not, see <http://www.gnu.org/licenses/>.

#pragma once

#include <boost/log/trivial.hpp>
#include <radiotray-ng/helpers.hpp>
#include <vector>
#include <string>
#include <iomanip>

using playlist_t = std::vector<std::string>;

// logging
#define LOG(x)                            BOOST_LOG_TRIVIAL(x) << "(" << radiotray_ng::basename(__FILE__) << ":"  << __LINE__ << ") - "

// event data keys
#define ERROR_KEY                         "error"
#define SONG_KEY                          "song"
#define STATE_KEY                         "state"
#define BOOKMARKS_KEY                     "bookmarks"
#define MESSAGE_KEY                       "message"

// state data values
#define STATE_BUFFERING                   "buffering"
#define STATE_PLAYING                     "playing"
#define STATE_STOPPED                     "stopped"
#define STATE_CONNECTING                  "connecting"

// tag data values
#define TAG_ARTIST                        "artist"
#define TAG_BITRATE                       "bitrate"
#define TAG_CODEC                         "audio codec"
#define TAG_TITLE                         "title"

// dbus json state message
#define DBUS_MSG_STATE_KEY                "state"
#define DBUS_MSG_VOLUME_KEY               "volume"
#define DBUS_MSG_TITLE_KEY                "title"
#define DBUS_MSG_ARTIST_KEY               "artist"
#define DBUS_MSG_STATION_KEY              "station"
#define DBUS_MSG_GROUP_KEY                "group"
#define DBUS_MSG_CODEC_KEY                "codec"
#define DBUS_MSG_BITRATE_KEY              "bitrate"
#define DBUS_MSG_IMAGE_KEY                "image"

// http codes
#define HTTP_OK                           (200)
#define HTTP_NOT_FOUND                    (404)

// config keys
#define BOOKMARKS_KEY                     "bookmarks"
#define BOOKMARK_EDITOR_KEY               "bookmark-editor"
#define BUFFER_SIZE_KEY                   "buffer-size"
#define BUFFER_DURATION_KEY               "buffer-duration"
#define COMPACT_MENU_KEY                  "compact-menu"
#define DEBUG_LOGGING_KEY                 "debug-logging"
#define FILE_MONITOR_KEY                  "file-monitor"
#define FILE_MONITOR_INTERVAL_KEY         "file-monitor-interval"
#define HTTP_TIMEOUT_KEY                  "http-timeout"
#define INVERT_MENU_KEY                   "invert-menu"
#define LAST_STATION_GROUP_KEY            "last-station-group"
#define LAST_STATION_KEY                  "last-station"
#define LAST_STATION_NOTIFICATION_KEY     "last-station-notifications"
#define MEDIA_KEY_MAPPING_KEY             "media-key-mapping"
#define MEDIA_KEY_NEXT_STAITON_KEY        "media-key-next-station"
#define MEDIA_KEY_PREVIOUS_STATION_KEY    "media-key-previous-station"
#define MEDIA_KEY_VOLUME_DOWN_KEY         "media-key-volume-down"
#define MEDIA_KEY_VOLUME_UP_KEY           "media-key-volume-up"
#define MEDIA_KEY_OLD_DBUS_NAME_KEY       "media-keys-old-dbus-name"
#define NOTIFICATION_KEY                  "notifications"
#define NOTIFICATION_VERBOSE_KEY          "notification-verbose"
#define RADIOTRAY_NG_ICON_ON_KEY          "radiotray-ng-on"
#define RADIOTRAY_NG_ICON_OFF_KEY         "radiotray-ng-off"
#define RADIOTRAY_NG_NOTIFICATION_KEY     "radiotray-ng-notification"
#define ROOT_GROUP_BOTTOM_POS_KEY         "root-group-bottom-pos"
#define SLEEP_TIMER_KEY                   "sleep-timer"
#define SPLIT_TITLE_KEY                   "split-title"
#define IHR_TITLE_KEY                     "ihr-title"
#define TAG_INFO_VERBOSE_KEY              "tag-info-verbose"
#define TRACK_INFO_COPY_KEY               "track-info-copy"
#define VOLUME_LEVEL_KEY                  "volume-level"
#define VOLUME_MAX_LEVEL_KEY              "volume-max-level"
#define VOLUME_STEP_KEY                   "volume-step"
#define WRAP_TRACK_INFO_KEY               "wrap-track-info"
#define WRAP_TRACK_INFO_LEN_KEY           "wrap-track-info-len"

// conf files etc.
#define RTNG_BOOKMARK_FILE                "bookmarks.json"
#define RTNG_CONFIG_FILE                  "radiotray-ng.json"
#define RTNG_DEFAULT_BOOKMARK_FILE        "/usr/share/radiotray-ng/bookmarks.json"
#define RTNG_DEFAULT_INSTALL_DIR          "/usr/share/radiotray-ng/"

// misc
#define ROOT_BOOKMARK_GROUP               "root"
#define APP_NAME_DISPLAY                  "Radiotray-NG"
#define APP_NAME                          "radiotray-ng"
#define APP_COPYRIGHT                     "Copyright © 2017-2019 - The Radiotray-NG Authors"
#define APP_AUTHOR                        "Edward G. Bruck"
#define APP_WEBSITE                       "https://www.github.com/ebruck/radiotray-ng"
#define RADIOTRAY_NG_LOGO_ICON            "radiotray-ng.png"

// defaults
#define DEFAULT_BUFFER_DURATION_VALUE            (2)
#define DEFAULT_BUFFER_SIZE_VALUE                uint32_t(320000)
#define DEFAULT_BOOKMARK_EDITOR                  "rtng-bookmark-editor"
#define DEFAULT_COMPACT_MENU_VALUE               (false)
#define DEFAULT_DEBUG_LOGGING_VALUE              (false)
#define DEFAULT_FILE_MONITOR_VALUE               (true)
#define DEFAULT_FILE_MONITOR_INTERVAL_VALUE      (10)
#define DEFAULT_HTTP_TIMEOUT_VALUE               (15)
#define DEFAULT_INVERT_MENU_VALUE                (false)
#define DEFAULT_MEDIA_KEY_MAPPING_VALUE          (false)
#define DEFAULT_MEDIA_KEY_NEXT_STATION_VALUE     "Next"
#define DEFAULT_MEDIA_KEY_PREVIOUS_STATION_VALUE "Previous"
#define DEFAULT_MEDIA_KEY_VOLUME_DOWN_VALUE      ""
#define DEFAULT_MEDIA_KEY_VOLUME_UP_VALUE        ""
#define DEFAULT_RADIOTRAY_NG_ICON_ON_VALUE       "radiotray-ng-on"
#define DEFAULT_RADIOTRAY_NG_ICON_OFF_VALUE      "radiotray-ng-off"
#define DEFAULT_RADIOTRAY_NG_NOTIFICATION_VALUE  "radiotray-ng-notification"
#define DEFAULT_ROOT_GROUP_BOTTOM_POS_VALUE      (true)
#define DEFAULT_MEDIA_KEY_OLD_DBUS_NAME_VALUE    (false)
#define DEFAULT_NOTIFICATION_VALUE               (true)
#define DEFAULT_NOTIFICATION_VERBOSE_VALUE       (true)
#define DEFAULT_SLEEP_TIMER_VALUE                uint32_t(60)
#define DEFAULT_SPLIT_TITLE_VALUE                (true)
#define DEFAULT_IHR_TITLE_KEY_VALUE              (true)
#define DEFAULT_STATION_IMAGE_VALUE              ""
#define DEFAULT_TAG_INFO_VERBOSE_VALUE           (false)
#define DEFAULT_STATION_NOTIFICATION_VALUE       (true)
#define DEFAULT_TRACK_INFO_COPY_VALUE            (false)
#define DEFAULT_VOLUME_LEVEL_MAX_VALUE           uint32_t(200)
#define DEFAULT_VOLUME_LEVEL_VALUE               uint32_t(100)
#define DEFAULT_VOLUME_STEP_VALUE                uint32_t(1)
#define DEFAULT_WRAP_TRACK_INFO_VALUE            (true)
#define DEFAULT_WRAP_TRACK_INFO_LEN_VALUE        (40)
